sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala

```diff
@@ -126,35 +126,32 @@ abstract class PartitioningAwareFileIndex(
     val caseInsensitiveOptions = CaseInsensitiveMap(parameters)
     val timeZoneId = caseInsensitiveOptions.get(DateTimeUtils.TIMEZONE_OPTION)
       .getOrElse(sparkSession.sessionState.conf.sessionLocalTimeZone)
-
-    userPartitionSchema match {
+    val inferredPartitionSpec = PartitioningUtils.parsePartitions(
+      leafDirs,
+      typeInference = sparkSession.sessionState.conf.partitionColumnTypeInferenceEnabled,
```

Comment:
this is causing a behavior change in Spark 2.4.0 reported in SPARK-26188. Why did we need this change?

Reply:
Before this patch, there was a subtle difference between with and without a user-provided partition schema:
1. with user-provided partition schema, we should not infer data types. We should infer as string and cast to user-provided type
2. without user-provided partition schema, we should infer the data type(with a config)

So it was wrong to unify these 2 code paths. @gengliangwang can you change it back?

Reply:
@mgaido91 Thanks for the investigation!!
I will fix it and add test case.

Reply:
actually the investigation was done by the reported of SPARK-26188, I did nothing... Thanks for doing that @gengliangwang and thanks for your comment @cloud-fan